src/Formats/SchemaInferenceUtils.cpp

```diff
@@ -67,24 +74,47 @@ namespace
             if (isNothing(type))
                 type = not_nothing_type;
         }
+
+        type_indexes.erase(TypeIndex::Nothing);
+    }
+
+    /// If we have both Int64 and UInt64, convert all Int64 to UInt64,
+    /// because UInt64 is inferred only in case of Int64 overflow.
+    void transformIntegers(DataTypes & data_types, TypeIndexesSet & type_indexes)
+    {
+        if (!type_indexes.contains(TypeIndex::Int64) || !type_indexes.contains(TypeIndex::UInt64))
+            return;
+
+        for (auto & type : data_types)
+        {
+            if (WhichDataType(type).isInt64())
+                type = std::make_shared<DataTypeUInt64>();
+        }
+
+        type_indexes.erase(TypeIndex::Int64);
     }
 
     /// If we have both Int64 and Float64 types, convert all Int64 to Float64.
-    void transformIntegersAndFloatsToFloats(DataTypes & data_types, const TypeIndexesSet & type_indexes)
+    void transformIntegersAndFloatsToFloats(DataTypes & data_types, TypeIndexesSet & type_indexes)
     {
-        if (!type_indexes.contains(TypeIndex::Int64) || !type_indexes.contains(TypeIndex::Float64))
+        bool have_floats = type_indexes.contains(TypeIndex::Float64);
+        bool have_integers = type_indexes.contains(TypeIndex::Int64) ||  type_indexes.contains(TypeIndex::UInt64);
+        if (!have_integers || !have_floats)
             return;
 
         for (auto & type : data_types)
         {
             if (isInteger(type))
```

Comment:
`isInteger()` will also return true for 8, 16 and 32-bit integers. It looks inconsistent with the above explicit checks for 64-integers (and removal of types). Is it okay?

Reply:
Other types just cannot appear in schema inference at all. But agree, let's check more explicitly